File: frontend-next/lib/chat/rag/retrieve.ts
import type { RagChunk } from "./types"
import { tokenizeForSearch } from "./tokenize"

function scoreChunk(queryTokens: string[], chunk: RagChunk): number {
  if (queryTokens.length === 0) return 0

  const haystack = `${chunk.title} ${chunk.text}`.toLowerCase()
  let score = 0

  for (const token of queryTokens) {
    if (haystack.includes(token)) {
      score += token.length >= 4 ? 3 : 2
    }
    if (chunk.title.toLowerCase().includes(token)) {
      score += 2
    }
  }

  return score
}

export function retrieveRagChunks(
  question: string,
  corpus: RagChunk[],
  topK = 8,
): RagChunk[] {
  const queryTokens = tokenizeForSearch(question)
  const minScore = queryTokens.length > 0 ? 2 : 0

  const ranked = corpus
    .map((chunk) => ({
      chunk,
      score: scoreChunk(queryTokens, chunk),
    }))
    .filter((item) => item.score >= minScore)
    .sort((a, b) => b.score - a.score)

  const selected =
    ranked.length > 0
      ? ranked.slice(0, topK)
      : corpus.slice(0, Math.min(topK, corpus.length)).map((chunk) => ({
          chunk,
          score: 0,
        }))

  return selected.map(({ chunk, score }) => ({
    ...chunk,
    score,
  }))
}

export function formatRagContext(chunks: RagChunk[]): string {
  if (chunks.length === 0) {
    return "=== RAG 검색 결과 ===\n(관련 문서를 찾지 못했습니다.)"
  }

  const body = chunks
    .map(
      (chunk, index) =>
        `[${index + 1}] (${chunk.source}) ${chunk.title}\n${chunk.text}`,
    )
    .join("\n\n")

  return `=== RAG 검색 결과 (상위 ${chunks.length}건) ===\n${body}`
}

File: backend/app/application/chat/rag/retrieve.py
from __future__ import annotations

from typing import Any

from app.application.chat.rag.corpus import tokenize_for_search


def _score_chunk(query_tokens: list[str], chunk: dict[str, Any]) -> int:
    if not query_tokens:
        return 0
    haystack = f"{chunk.get('title', '')} {chunk.get('text', '')}".lower()
    score = 0
    for token in query_tokens:
        if token in haystack:
            score += 3 if len(token) >= 4 else 2
        if token in chunk.get("title", "").lower():
            score += 2
    return score


def retrieve_rag_chunks(
    question: str,
    corpus: list[dict[str, Any]],
    top_k: int = 8,
) -> list[dict[str, Any]]:
    query_tokens = tokenize_for_search(question)
    min_score = 2 if query_tokens else 0

    ranked = sorted(
        (
            {"chunk": c, "score": _score_chunk(query_tokens, c)}
            for c in corpus
        ),
        key=lambda item: item["score"],
        reverse=True,
    )
    ranked = [item for item in ranked if item["score"] >= min_score]

    selected = ranked[:top_k] if ranked else [
        {"chunk": c, "score": 0} for c in corpus[: min(top_k, len(corpus))]
    ]

    return [{**item["chunk"], "score": item["score"]} for item in selected]


def format_rag_context(chunks: list[dict[str, Any]]) -> str:
    if not chunks:
        return "=== RAG 검색 결과 ===\n(관련 문서를 찾지 못했습니다.)"
    body = "\n\n".join(
        f"[{i + 1}] ({c.get('source')}) {c.get('title')}\n{c.get('text')}"
        for i, c in enumerate(chunks)
    )
    return f"=== RAG 검색 결과 (상위 {len(chunks)}건) ===\n{body}"
